### PR TITLE
Vision template

### PR DESCRIPTION
**Pull request by EliyaFishman**

### CODE REVIEW

File: src/main/java/frc/robot/vision/Target.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package frc.robot.vision;
+
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Target.java]: just move the enum to a new folder for enums

File: src/main/java/frc/robot/vision/Limelight.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    setPipeline(target.getIndex());
+  }
+
+  /**
+   * @return the vector between the middle of the robot and and the target.
+   */
+  private Vector2d calculateVector() {
+    //This is the offset vector.
+    Vector2d MiddleToLimelight = new Vector2d(-robotConstants.visionConstants.LIMELIGHT_OFFSET_X, -robotConstants.visionConstants.LIMELIGHT_OFFSET_Y);
+    //This is the the vector from the limelight to the target.
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Limelight.java:204]: Nice calculation!
change MiddleToLimelight TO middleToLimelight pls

File: src/main/java/frc/robot/vision/Limelight.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    CamMode(int value) {
+      this.value = value;
+    }
+
+    public int getValue() {
+      return value;
+    }
+
+  }
+
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Limelight.java]: move CamMode to a new folder, call it:
enums
and remove the blank line here:
  }
//here
  }

File: src/main/java/frc/robot/vision/Limelight.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+  /**
+   * @return The distance between the the target and the limelight
+   */
+  //TODO: set real function
+  public double getDistanceFromLimelight() {
+    double x = getTy();
+    return robotConstants.visionConstants.DISTANCE_CALCULATION_A_COEFFICIENT * x * x +
+            robotConstants.visionConstants.DISTANCE_CALCULATION_B_COEFFICIENT * x;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Limelight.java:81]: use Math.pow(x, 2); 
I think its more readable

File: src/main/java/frc/robot/vision/Limelight.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+    LedMode(int value) {
+      this.value = value;
+    }
+
+    public int getValue() {
+      return value;
+    }
+  }
+
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Limelight.java]: also, this enum. 
everyone in its own class

File: src/main/java/frc/robot/vision/Limelight.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+  private Vector2d calculateVector() {
+    //This is the offset vector.
+    Vector2d MiddleToLimelight = new Vector2d(-robotConstants.visionConstants.LIMELIGHT_OFFSET_X, -robotConstants.visionConstants.LIMELIGHT_OFFSET_Y);
+    //This is the the vector from the limelight to the target.
+    Vector2d limelightToTarget = new Vector2d(0, getDistanceFromLimelight());
+    limelightToTarget.rotate(getTx());
+    // The offset is subtracted from the limelightToTarget vector in order to get the final vector.
+    return new Vector2d(MiddleToLimelight.x + limelightToTarget.x, MiddleToLimelight.y + limelightToTarget.y);
+  }
+}
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Limelight.java:210]: from the limelight api docs:
snapshot | Allows users to take snapshots during a match
-- | --
0 | Stop taking snapshots
1 | Take two snapshots per second

what do you say? lets add the option of taking snapshots for pipeline calibration if we would not have time in the competition morning to calibrate (bad, i know).

File: src/main/java/frc/robot/vision/Limelight.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    }
+  }
+
+  public void toggleLedMode() {
+    if (getLedMode().equals(LedMode.off))
+      setLedMode(LedMode.on);
+    else
+      setLedMode(LedMode.off);
+
+  }
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Limelight.java:126]: setLedMode(LedMode.off);
//remove blank line
  }

File: src/main/java/frc/robot/vision/Limelight.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+
+  /**
+   * @return the current target in the NetworkTable.
+   */
+  public Target getTarget() {
+    int index = (int) pipeline.getDouble(0);
+    return Target.values()[index];
+  }
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Limelight.java:157]: Target by it's pipeline... but what happens if you have two targets that work with the same pipeline?

File: src/main/java/frc/robot/vision/Limelight.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+  /**
+   * @param pipeline pipeline index to be changed to.
+   */
+  public void setPipeline(int pipeline) {
+    this.pipeline.setNumber(pipeline);
+  }
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Limelight.java:165]: flush memory here?
 NetworkTableInstance.getDefault().flush();

File: src/main/java/frc/robot/vision/Target.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  Target(int index) {
+    this.index = index;
+
+  }
+
+  public int getIndex() {
+    return index;
+  }
+
+}
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Target.java]: remove blank lines: 

Target(int index) {
    this.index = index;
//here
  }

  public int getIndex() {
    return index;
  }
//here
}

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 11]
```diff
@@ -20,8 +21,18 @@
...
+        public PIDSettings visionDistanceSettings;
     }
 
-    // More static class here! 
+    public static class VisionConstants {
+        public double DISTANCE_CALCULATION_A_COEFFICIENT;
+        public double DISTANCE_CALCULATION_B_COEFFICIENT;
+        public double LIMELIGHT_OFFSET_X;
+        public double LIMELIGHT_OFFSET_Y;
+        public double TARGET_NOT_FOUND_WAIT_TIME;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/constants/RobotConstants.java:34]: visionConstants.LIMELIGHT_OFFSET_X = 0;
visionConstants.LIMELIGHT_OFFSET_Y = 0;
//add commant explaining its masured from the middle of the robot...

File: src/main/java/frc/robot/utils/Logger.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    private StringBuilder data;
+    /**
+     * @param name the name of the file. Should end with .csv
+     * @param columns the columns name (ex. velocity, acceleration)
+     */
+    public Logger(String name, String... columns) {
+        if(!name.endsWith(".csv"))
+            name = name+".csv";
+        this.path =  Filesystem.getOperatingDirectory()+"/logs/"+ getTimeStamp() + '-' + name;
+        data = new StringBuilder();
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/utils/Logger.java:31]: this.path =  Filesystem.getOperatingDirectory()+"/logs/"+ getTimeStamp() + '-' + name;
TO
path = Filesystem.getOperatingDirectory()+"/logs/"+ getTimeStamp() + '-' + name;
important - look that you have between this.path =  Filesystem two spaces, and you need only one

File: src/main/java/frc/robot/utils/Logger.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        for (int i = 0; i < values.length; i++) {
+            array[i] = Double.toString(values[i]);
+        }
+        log(array);
+    }
+    /**
+     * Saves the file and closes it. Should only be called once.
+     */
+    public void close(){
+        File file = new File(path);
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/utils/Logger.java:61]: public void close(){ 
TO
public void close() {

File: src/main/java/frc/robot/utils/Logger.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+            file.createNewFile();
+            FileWriter writer = new FileWriter(file);
+            writer.write(data.toString());
+            writer.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+    }
+
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/utils/Logger.java:75]: e.printStackTrace();
        }
//remove blank space 
    }

File: src/main/java/frc/robot/utils/Logger.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    private String getTimeStamp(){
+        Timestamp timestamp = new Timestamp(System.currentTimeMillis());
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/utils/Logger.java:77]: private String getTimeStamp(){
TO 
  private String getTimeStamp() {

File: src/main/java/frc/robot/vision/Limelight.java [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+   */
+  public double getDistance() {
+    return calculateVector().magnitude();
+  }
+
+  /**
+   * @return the angle from the middle of the robot to the target
+   */
+  public double getAngle(){
+    Vector2d vector = calculateVector();
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Limelight.java:95]: public double getAngle(){ 
TO
 public double getAngle() {

File: src/main/java/frc/robot/utils/Logger.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        if(!name.endsWith(".csv"))
+            name = name+".csv";
+        this.path =  Filesystem.getOperatingDirectory()+"/logs/"+ getTimeStamp() + '-' + name;
+        data = new StringBuilder();
+        log(columns);
+    }
+
+    /**
+     * @param values the values to be added to the logger
+     */
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/utils/Logger.java:37]: values the values to be added to the logger
also, explain that it creating the columns in the CSV file

File: src/main/java/frc/robot/utils/Logger.java [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    public void log(double ... values) {
+        String[] array = new String[values.length];
+        for (int i = 0; i < values.length; i++) {
+            array[i] = Double.toString(values[i]);
+        }
+        log(array);
+    }
+    /**
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/utils/Logger.java:57]: }
    /**
add space between them

File: src/main/java/frc/robot/utils/Logger.java [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+     * Saves the file and closes it. Should only be called once.
+     */
+    public void close(){
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/utils/Logger.java]: */
    public void close(){
add space between them

File: src/main/java/frc/robot/utils/Logger.java [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+    /**
+     * @param values the values to be added to the logger
+     */
+    public void log(String... values) {
+        for (String value : values) {
+            data.append(value).append(',');
+        }
+        data.setCharAt(data.length() - 1, '\n');
+    }
+    /**
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/utils/Logger.java:45]: }
    /**
add space between them

File: src/main/java/frc/robot/utils/Logger.java [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.sql.Timestamp;
+
+public class Logger {
+    private String path;
+    private StringBuilder data;
+    /**
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/utils/Logger.java:23]: private StringBuilder data;
    /**
add space between them:
   private StringBuilder data;

   /**

File: src/main/java/frc/robot/utils/Logger.java [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+     * @param name the name of the file. Should end with .csv
+     * @param columns the columns name (ex. velocity, acceleration)
+     */
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/utils/Logger.java:26]: add to the doc a short explanation about what the logger does...

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java [hunk 23]
```diff
@@ -0,0 +1,69 @@
+package frc.robot.vision;
+
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.Robot;
+import frc.robot.utils.Logger;
+
+import java.util.function.Supplier;
+
+
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/CalibrateVisionDistance.java:9]: too many spaces

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java [hunk 24]
```diff
@@ -0,0 +1,69 @@
+package frc.robot.vision;
+
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/CalibrateVisionDistance.java:3]: too many spaces

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+import frc.robot.Robot;
+import frc.robot.utils.Logger;
+
+import java.util.function.Supplier;
+
+
+public class CalibrateVisionDistance extends CommandBase {
+  private static final int DEFAULT_DELTA_DISTANCE = 30;
+  private static final int AMOUNT_OF_LOGS = 8;
+  private double currentDistance = 0;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/CalibrateVisionDistance.java:13]: private static final int AMOUNT_OF_LOGS = 8; 
rise to 12...
AND change to DEFAULT_AMOUNT_OF_LOGS

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+  private boolean isPressed = false;
+  private Supplier<Boolean> logButton;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/CalibrateVisionDistance.java:15]: isPressed = false; move to the initialize, no?

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  /**
+   * @param logButton whenever the supplier toggles to true - log the values.
+   * @param deltaDistance the distance between each log.
+   */
+  public CalibrateVisionDistance(Supplier<Boolean> logButton, double deltaDistance) {
+    this.logButton = logButton;
+    this.deltaDistance = deltaDistance;
+  }
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/CalibrateVisionDistance.java:45]: add a constructor with - int amountOfLogs and default of DEFAULT_AMOUNT_OF_LOGS

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+  public void execute() {
+    if (logButton.get()) {
+      if (!isPressed) {
+        isPressed = true;
+        logger.log(Robot.limelight.getTy(), currentDistance);
+        currentDistance += deltaDistance;
+      }
+    } else
+      isPressed = false;
+  }
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/CalibrateVisionDistance.java:65]: to make it more precise I want you to add another value to the log, of the average distance measured by the two encoders of the drivetrain **AND don't forget to reset them in the initialize!!!**

File: src/main/java/frc/robot/vision/TurnToTarget.java [hunk 29]
```diff
@@ -0,0 +1,78 @@
...
+  private PIDController rotationPIDController;
+  private double lastTimeSeenTarget;
+
+  /**
+   * @param target The target the robot will follow
+   * @param output accepts rotation output.
+   */
+  public TurnToTarget(Target target, DoubleConsumer output) {
+    addRequirements(Robot.drivetrain);
+    this.target = target;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/TurnToTarget.java:32]: addRequirements(Robot.drivetrain); what if it a shooter subsystem? add the subsystem to require in the constructor
AND another idea is to make an interface for all of our subsystems that will contain a moveWithVision func or something like that and then your output would be inserted to this func and all you would need to use this command is to give it a subsystem and target... what do you say?

File: src/main/java/frc/robot/vision/TurnToTarget.java [hunk 30]
```diff
@@ -0,0 +1,78 @@
...
+    return ((Timer.getFPGATimestamp() - lastTimeSeenTarget) > robotConstants.visionConstants.TARGET_NOT_FOUND_WAIT_TIME)
+            || rotationPIDController.atSetpoint();
+  }
+
+  @Override
+  public void end(boolean interrupted) {
+    output.accept(0.0);
+    limelight.setLedMode(LedMode.off);
+    limelight.setCamMode(CamMode.driver);
+  }
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/TurnToTarget.java:67]: limelight.setLedMode(LedMode.off);
    limelight.setCamMode(CamMode.driver);

what do you say about making it one void func called closeVision (or something like that) in the Limelight class...?

File: src/main/java/frc/robot/vision/TurnToTarget.java [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+
+  @Override
+  public void initialize() {
+    rotationPIDController.reset();
+    lastTimeSeenTarget = Timer.getFPGATimestamp();
+    // Configure the limelight to start computing vision.
+    limelight.setPipeline(target);
+    limelight.setCamMode(CamMode.vision);
+    limelight.setLedMode(LedMode.on);
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/TurnToTarget.java]: limelight.setPipeline(target);
    limelight.setCamMode(CamMode.vision);
    limelight.setLedMode(LedMode.on);

what do you say about making it one void func called startVision(Target target) (or something like that) in the Limelight class...?

File: src/main/java/frc/robot/vision/TurnToTarget.java [hunk 32]
```diff
@@ -0,0 +1,78 @@
...
+  public void end(boolean interrupted) {
+    output.accept(0.0);
+    limelight.setLedMode(LedMode.off);
+    limelight.setCamMode(CamMode.driver);
+  }
+
+  public void enableTuning() {
+    SmartDashboard.putData("PID/visionRotation", rotationPIDController);
+  }
+}
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/TurnToTarget.java:72]: SmartDashboard.putData("PID/visionRotation", rotationPIDController);
Very nice :)

File: src/main/java/frc/robot/vision/TurnToTarget.java [hunk 33]
```diff
@@ -0,0 +1,78 @@
...
+    limelight.setPipeline(target);
+    limelight.setCamMode(CamMode.vision);
+    limelight.setLedMode(LedMode.on);
+  }
+
+  @Override
+  public void execute() {
+    if (limelight.getTv()) {
+      double rotationOutput = rotationPIDController.calculate(limelight.getAngle());
+      output.accept(rotationOutput);
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/TurnToTarget.java]: double rotationOutput = rotationPIDController.calculate(limelight.getAngle());
      output.accept(rotationOutput);

I think you can just make it:
      output.accept(rotationPIDController.calculate(limelight.getAngle()));

File: src/main/java/frc/robot/vision/FollowTarget.java [hunk 34]
```diff
@@ -0,0 +1,86 @@
+package frc.robot.vision;
+
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/FollowTarget.java:2]: just see my comments from the TurnToTarget command it is basically the same...

**Review by HilelAmiel — CHANGES_REQUESTED** [review]
good job, fix and return to me when you finish

File: src/main/java/frc/robot/vision/Limelight.java (hunk 3 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java]: what about LedMode?

File: src/main/java/frc/robot/vision/Limelight.java (hunk 6 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java:210]: good idea

File: src/main/java/frc/robot/vision/Limelight.java (hunk 8 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java:157]: I don't think we even use this function.
I will change it to return the pipeline

File: src/main/java/frc/robot/constants/RobotConstants.java (hunk 11 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/constants/RobotConstants.java:34]: It is set to 0 in RobotA.
I thought we didn't want to initialize constants in RobotConstants?

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java (hunk 26 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/CalibrateVisionDistance.java:15]: Yes you are correct

File: src/main/java/frc/robot/vision/TurnToTarget.java (hunk 29 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/TurnToTarget.java:32]: I think the first idea is good.

Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/TurnToTarget.java:32]: I just copied the FollowTarget class and forgot to change this

File: src/main/java/frc/robot/vision/TurnToTarget.java (hunk 30 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/TurnToTarget.java:67]: good idea

File: src/main/java/frc/robot/vision/TurnToTarget.java (hunk 31 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/TurnToTarget.java]: ok

File: src/main/java/frc/robot/vision/FollowTarget.java (hunk 34 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/FollowTarget.java:2]: Should I inline distanceOutput and turnOutput in FollowTarget too? It becomes kinda long:
`      output.accept(rotationPIDController.calculate(limelight.getAngle()), distancePIDController.calculate(limelight.getDistance()));`

Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/FollowTarget.java:2]: and about the subsystem not being necessarily drivetrain, there is any other subsystem that can move in such a way?

File: src/main/java/frc/robot/vision/Target.java (hunk 1 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Target.java]: fixed

File: src/main/java/frc/robot/vision/Limelight.java (hunk 2 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java:204]: fixed

File: src/main/java/frc/robot/vision/Limelight.java (hunk 3 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java]: fixed

File: src/main/java/frc/robot/vision/Limelight.java (hunk 4 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java:81]: fixed

File: src/main/java/frc/robot/vision/Limelight.java (hunk 5 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java]: fixed

File: src/main/java/frc/robot/vision/Limelight.java (hunk 6 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java:210]: added

File: src/main/java/frc/robot/vision/Limelight.java (hunk 7 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java:126]: fixed

File: src/main/java/frc/robot/vision/Limelight.java (hunk 8 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java:157]: changed

File: src/main/java/frc/robot/vision/Limelight.java (hunk 9 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java:165]: fixed

File: src/main/java/frc/robot/vision/Target.java (hunk 10 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Target.java]: fixed

File: src/main/java/frc/robot/utils/Logger.java (hunk 12 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/utils/Logger.java:31]: fixed

File: src/main/java/frc/robot/utils/Logger.java (hunk 13 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/utils/Logger.java:61]: fixed

File: src/main/java/frc/robot/utils/Logger.java (hunk 14 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/utils/Logger.java:75]: fixed

File: src/main/java/frc/robot/utils/Logger.java (hunk 15 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/utils/Logger.java:77]: fixed

File: src/main/java/frc/robot/vision/Limelight.java (hunk 16 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java:95]: fixed

File: src/main/java/frc/robot/utils/Logger.java (hunk 17 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/utils/Logger.java:37]: fixed

File: src/main/java/frc/robot/utils/Logger.java (hunk 18 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/utils/Logger.java:57]: fixed

File: src/main/java/frc/robot/utils/Logger.java (hunk 19 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/utils/Logger.java]: fixed

File: src/main/java/frc/robot/utils/Logger.java (hunk 20 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/utils/Logger.java:45]: fixed

File: src/main/java/frc/robot/utils/Logger.java (hunk 21 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/utils/Logger.java:23]: fixed

File: src/main/java/frc/robot/utils/Logger.java (hunk 22 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/utils/Logger.java:26]: fixed

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java (hunk 23 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/CalibrateVisionDistance.java:9]: fixed

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java (hunk 24 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/CalibrateVisionDistance.java:3]: fixed

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java (hunk 25 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/CalibrateVisionDistance.java:13]: fixed

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java (hunk 26 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/CalibrateVisionDistance.java:15]: fixed

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java (hunk 27 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/CalibrateVisionDistance.java:45]: fixed

File: src/main/java/frc/robot/vision/CalibrateVisionDistance.java (hunk 28 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/CalibrateVisionDistance.java:65]: fixed

File: src/main/java/frc/robot/vision/TurnToTarget.java (hunk 29 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/TurnToTarget.java:32]: fixed

File: src/main/java/frc/robot/vision/TurnToTarget.java (hunk 30 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/TurnToTarget.java:67]: done

File: src/main/java/frc/robot/vision/TurnToTarget.java (hunk 31 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/TurnToTarget.java]: done

File: src/main/java/frc/robot/vision/TurnToTarget.java (hunk 32 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/TurnToTarget.java:72]: Thank you

File: src/main/java/frc/robot/vision/TurnToTarget.java (hunk 33 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/TurnToTarget.java]: done

File: src/main/java/frc/robot/constants/RobotConstants.java (hunk 11 above)
Comment (HilelAmiel) [reply src/main/java/frc/robot/constants/RobotConstants.java:34]: ah... no, its my mistake i just copied the comment in robotA class you did well.

File: src/main/java/frc/robot/enums/CamMode.java [hunk 35]
```diff
@@ -0,0 +1,19 @@
...
+
+  CamMode(int value) {
+    this.value = value;
+  }
+
+  public int getValue() {
+    return value;
+  }
+
+}
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/enums/CamMode.java:18]: return value;
  }
//here
}

delete blank space

File: src/main/java/frc/robot/vision/Limelight.java [hunk 36]
```diff
@@ -0,0 +1,210 @@
...
+  public Limelight() {
+    this(DEFAULT_TABLE_KEY);
+  }
+
+  public void startVision(Target target) {
+    setPipeline(target);
+    setCamMode(CamMode.vision);
+    setLedMode(LedMode.on);
+  }
+
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/vision/Limelight.java]: move startVision() after or before stopVision() down here...

File: src/main/java/frc/robot/vision/FollowTarget.java (hunk 34 above)
Comment (HilelAmiel) [reply src/main/java/frc/robot/vision/FollowTarget.java:2]: no, you right you can leave the drivetrain here as it is

Comment (HilelAmiel) [reply src/main/java/frc/robot/vision/FollowTarget.java:2]: > Should I inline distanceOutput and turnOutput in FollowTarget too? It becomes kinda long:
> ` output.accept(rotationPIDController.calculate(limelight.getAngle()), distancePIDController.calculate(limelight.getDistance()));`

you right also here. these variables should stay.

**Review by HilelAmiel — CHANGES_REQUESTED** [review]
last changes

File: src/main/java/frc/robot/vision/Limelight.java (hunk 36 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/vision/Limelight.java]: done

File: src/main/java/frc/robot/enums/CamMode.java (hunk 35 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/enums/CamMode.java:18]: done

**Review by HilelAmiel — APPROVED** [review]
Done! Now we have Vision.

**Pull request merged**